### ISSUE TITLE
System background jobs run twice

### ISSUE THREAD

**Issue by emanuil-tolev** [DOAJ/doaj#1248]
This is not a huge problem, but it does mean we don't keep 2 days worth of sitemaps, just 1 day (since both copies we keep are from 1 day).

1. https://doaj.org/admin/background_jobs?source=%7B%22query%22%3A%7B%22filtered%22%3A%7B%22filter%22%3A%7B%22bool%22%3A%7B%22must%22%3A%5B%7B%22term%22%3A%7B%22action.exact%22%3A%22sitemap%22%7D%7D%5D%7D%7D%2C%22query%22%3A%7B%22query_string%22%3A%7B%22query%22%3A%22eadbb134a14d4a7ba87816e2cda18c7e%22%2C%22default_operator%22%3A%22AND%22%7D%7D%7D%7D%2C%22sort%22%3A%5B%7B%22created_date%22%3A%7B%22order%22%3A%22desc%22%7D%7D%5D%2C%22from%22%3A0%2C%22size%22%3A25%7D

2. https://doaj.org/admin/background_jobs?source=%7B%22query%22%3A%7B%22filtered%22%3A%7B%22filter%22%3A%7B%22bool%22%3A%7B%22must%22%3A%5B%7B%22term%22%3A%7B%22action.exact%22%3A%22sitemap%22%7D%7D%5D%7D%7D%2C%22query%22%3A%7B%22query_string%22%3A%7B%22query%22%3A%22f898a340b6ca48e89ee4c8fc7252620f%22%2C%22default_operator%22%3A%22AND%22%7D%7D%7D%7D%2C%22sort%22%3A%5B%7B%22created_date%22%3A%7B%22order%22%3A%22desc%22%7D%7D%5D%2C%22from%22%3A0%2C%22size%22%3A25%7D

None of the crontabs or anacrontabs on the gateway, the application, or the background application servers call the sitemap. The jobs must have been generated by huey.

huey-long-running does not mention sitemap in its log.
huey-main's log says:
```
[2017-03-22 08:00:12,476] INFO:huey.consumer.Scheduler:Scheduler:Scheduling periodic task queuecmd_scheduled_sitemap: queuecmd_scheduled_sitemap.
[2017-03-22 08:00:12,478] DEBUG:huey.consumer.Scheduler:Scheduler:Enqueued task: queuecmd_scheduled_sitemap: queuecmd_scheduled_sitemap
[2017-03-22 08:00:27,493] INFO:huey.consumer.Scheduler:Scheduler:Scheduling queuecmd_generate_sitemap: d3fe2333-2227-4e27-8a2d-7421183656e8 @2017-03-22 08:00:27 for execution
[2017-03-22 08:00:27,494] DEBUG:huey.consumer.Scheduler:Scheduler:Enqueued task: queuecmd_generate_sitemap: d3fe2333-2227-4e27-8a2d-7421183656e8 @2017-03-22 08:00:27
[2017-03-22 08:00:50,769] INFO:huey.consumer.Worker:Worker-1:Executing queuecmd_scheduled_sitemap: queuecmd_scheduled_sitemap
[2017-03-22 08:00:50,792] DEBUG:huey.consumer.Worker:Worker-1:Task queuecmd_scheduled_sitemap: queuecmd_scheduled_sitemap ran in 0.022s
[2017-03-22 08:00:50,794] INFO:huey.consumer.Worker:Worker-1:Adding queuecmd_generate_sitemap: 790e804d-98e1-467b-ba71-45fec03fe212 @2017-03-22 08:01:00 to schedule
[2017-03-22 08:01:00,531] INFO:huey.consumer.Scheduler:Scheduler:Scheduling queuecmd_generate_sitemap: 790e804d-98e1-467b-ba71-45fec03fe212 @2017-03-22 08:01:00 for execution
[2017-03-22 08:01:00,532] DEBUG:huey.consumer.Scheduler:Scheduler:Enqueued task: queuecmd_generate_sitemap: 790e804d-98e1-467b-ba71-45fec03fe212 @2017-03-22 08:01:00
[2017-03-22 08:01:01,109] INFO:huey.consumer.Worker:Worker-1:Executing queuecmd_generate_sitemap: 790e804d-98e1-467b-ba71-45fec03fe212 @2017-03-22 08:01:00
[2017-03-22 08:06:34,809] DEBUG:huey.consumer.Worker:Worker-1:Task queuecmd_generate_sitemap: 790e804d-98e1-467b-ba71-45fec03fe212 @2017-03-22 08:01:00 ran in 333.699s
```

Specifically:

```
[2017-03-22 08:00:27,493] INFO:huey.consumer.Scheduler:Scheduler:Scheduling queuecmd_generate_sitemap: d3fe2333-2227-4e27-8a2d-7421183656e8 @2017-03-22 08:00:27 for execution
[2017-03-22 08:01:00,531] INFO:huey.consumer.Scheduler:Scheduler:Scheduling queuecmd_generate_sitemap: 790e804d-98e1-467b-ba71-45fec03fe212 @2017-03-22 08:01:00 for execution
```

Looks like the job is scheduled twice. There are no superfluous huey processes running however. Only `main` and `long-running` are running, and just one of each. As I said above, those two lines are both in huey-main's log anyway.

```
cloo     23347  1170  6 Mar21 ?        00:48:27 /home/cloo/repl/production/doaj/bin/python /home/cloo/repl/production/doaj/bin/huey_consumer.py -v portality.tasks.consumer_main_queue.main_queue
cloo     23358  1170  0 Mar21 ?        00:03:21 /home/cloo/repl/production/doaj/bin/python /home/cloo/repl/production/doaj/bin/huey_consumer.py -v portality.tasks.consumer_long_running.long_running
```

So, not sure what's going on. Everything seems to be running correctly, but we got the job twice.

This is not a big problem at all, but we might learn something we don't know that would help troubleshoot other more pressing problems later (if any occur).

**Comment by emanuil-tolev** [comment]
@Steven-Eardley @richard-jones looks like all jobs triggered by a schedule, rather than a user, run twice. (= are triggered twice).

Here's some read_news and journal_csv duplicates in addition to the sitemap described in detail above.
![image](https://cloud.githubusercontent.com/assets/1190172/24193857/a06d0bae-0eeb-11e7-9a6e-697039b391ee.png)

**Comment by richard-jones** [comment]
Hm, very odd.  And this didn't happen on the test server?

It shouldn't cause any issues, other than unnecessary compute time, but it's obviously not right so would be useful to get to the bottom of it.

The mechanism by which scheduled tasks are done is as follows:

1. A scheduled task is added to the huey thread at start up, with a configured frequency
2. When the scheduled task is executed, it creates a job and adds it to the processing queue
3. The job is executed as normal

What this means is that the scheduled task does not actually run the job, it is only responsible for putting the job on the queue. 

Looking at the logs above, this makes sense with the following sections:

    [2017-03-22 08:00:12,476] INFO:huey.consumer.Scheduler:Scheduler:Scheduling periodic task queuecmd_scheduled_sitemap: queuecmd_scheduled_sitemap.
    [2017-03-22 08:00:12,478] DEBUG:huey.consumer.Scheduler:Scheduler:Enqueued task: queuecmd_scheduled_sitemap: queuecmd_scheduled_sitemap
    [2017-03-22 08:00:27,493] INFO:huey.consumer.Scheduler:Scheduler:Scheduling queuecmd_generate_sitemap: d3fe2333-2227-4e27-8a2d-7421183656e8 @2017-03-22 08:00:27 for execution
    [2017-03-22 08:00:27,494] DEBUG:huey.consumer.Scheduler:Scheduler:Enqueued task: queuecmd_generate_sitemap: d3fe2333-2227-4e27-8a2d-7421183656e8 @2017-03-22 08:00:27


It seems that huey itself is triggering the scheduled task twice, which I don't think is a problem with our code.  Instead it could be a problem with our implementation of the crontab function?

https://github.com/DOAJ/doaj/blob/develop/portality/tasks/redis_huey.py#L8

I did find a bug in the huey function when I did my original investigation, so maybe there's a problem there.
